### PR TITLE
IEP-690: Add an option to delete all configurations related to the project being deleted.

### PR DESCRIPTION
**Pull request by sigmaaa**
Added an option to delete all configurations related to the project being deleted. 

To verify create a few projects and multiple configurations to these projects
IEP-690

Right-click on the project in project explorer-> Delete

IEP-690-1

after the deletion
IEP-690-2

### CODE REVIEW

**Review by alirana01 — COMMENTED** [review]
About this
bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/dialogs/DeleteResourceWizard.java

Is this some internal file and do we need all of it or cant we just extend from the original class and simply override the performFinish method

Also we can write a functional test for this as well

**Comment by sigmaaa** [comment]
> About this bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/dialogs/DeleteResourceWizard.java
> 
> Is this some internal file and do we need all of it or cant we just extend from the original class and simply override the performFinish method
> 
> Also we can write a functional test for this as well

Hi @alirana01, thanks for the review. I tried to make it simple but the problem is that the UI is written in a private class "DeleteResourcesRefactoringConfigurationPage" and I wanted to add a new checkbox below the old one (Delete project contents on disk) and I didn't find a way to do it without overriding the whole private class

**Comment by kolipakakondal** [comment]
Here is the error message which is appearing while I click on the Delete menu option. Could you check how to avoid the Conflicting handlers issue?
 
```
!ENTRY org.eclipse.ui 4 4 2022-05-06 09:10:47.904
!MESSAGE Conflicting handlers for org.eclipse.ltk.ui.refactoring.commands.deleteResources: {com.espressif.idf.ui.handlers.DeleteResourceHandler@721af8d0} vs {org.eclipse.ltk.internal.ui.refactoring.actions.DeleteResourcesHandler} in: snapshotContext
!STACK 0
java.lang.IllegalStateException: Conflicting handlers: {[com.espressif.idf.ui.handlers.DeleteResourceHandler@721af8d0,  for 'org.eclipse.ltk.ui.refactoring.commands.deleteResources',  in WorkbenchContext, active=true, sourcePriority=0, participating=true]} vs {[org.eclipse.ltk.internal.ui.refactoring.actions.DeleteResourcesHandler,  for 'org.eclipse.ltk.ui.refactoring.commands.deleteResources',  in WorkbenchContext, active=true, sourcePriority=0, participating=true]} in: snapshotContext
	at org.eclipse.ui.internal.handlers.LegacyHandlerService$HandlerSelectionFunction.compute(LegacyHandlerService.java:145)
	at org.eclipse.e4.core.internal.contexts.ValueComputation.get(ValueComputation.java:70)
	at org.eclipse.e4.core.internal.contexts.EclipseContext.internalGet(EclipseContext.java:271)
	at org.eclipse.e4.core.internal.contexts.EclipseContext.internalGet(EclipseContext.java:282)
	at org.eclipse.e4.core.internal.contexts.EclipseContext.internalGet(EclipseContext.java:282)
	at org.eclipse.e4.core.internal.contexts.EclipseContext.internalGet(EclipseContext.java:282)
	at org.eclipse.e4.core.internal.contexts.EclipseContext.internalGet(EclipseContext.java:282)
	at org.eclipse.e4.core.internal.contexts.EclipseContext.internalGet(EclipseContext.java:282)
	at org.eclipse.e4.core.internal.contexts.EclipseContext.get(EclipseContext.java:237)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.lookUpHandler(HandlerServiceImpl.java:103)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.isHandled(HandlerServiceHandler.java:126)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:476)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:487)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:213)
	at org.eclipse.ui.internal.handlers.LegacyHandlerService.executeCommandInContext(LegacyHandlerService.java:440)
	at org.eclipse.ui.internal.ide.actions.LTKLauncher.runCommand(LTKLauncher.java:140)
	at org.eclipse.ui.internal.ide.actions.LTKLauncher.openDeleteWizard(LTKLauncher.java:57)
	at org.eclipse.ui.actions.DeleteResourceAction.run(DeleteResourceAction.java:452)
	at org.eclipse.ui.actions.BaseSelectionListenerAction.runWithEvent(BaseSelectionListenerAction.java:171)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4524)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1545)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1530)
	at org.eclipse.swt.widgets.Widget.notifyListeners(Widget.java:1324)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4310)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3883)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1041)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)

```

**Comment by kolipakakondal** [comment]
Created a new launch configuration using the toolbar 
New Launch Configuration > Run > ESP-IDF Application > Give some launch config name > Click finish > from the toolbar click on the gear icon next to the new launch configuration. Here is the error reported on the console and wizard didn't prompt
```

!ENTRY org.eclipse.ui 4 0 2022-05-06 09:15:29.921
!MESSAGE Unhandled event loop exception
!STACK 0
java.lang.NullPointerException
	at org.eclipse.cdt.launch.ui.corebuild.CoreBuildTab.getProject(CoreBuildTab.java:109)
	at org.eclipse.cdt.launch.ui.corebuild.CoreBuildTab.initializeFrom(CoreBuildTab.java:66)
	at org.eclipse.debug.ui.AbstractLaunchConfigurationTabGroup.initializeFrom(AbstractLaunchConfigurationTabGroup.java:93)
	at org.eclipse.launchbar.ui.internal.LaunchBarLaunchConfigDialog.createDialogArea(LaunchBarLaunchConfigDialog.java:165)
	at org.eclipse.jface.dialogs.TitleAreaDialog.createContents(TitleAreaDialog.java:162)
	at org.eclipse.jface.window.Window.create(Window.java:431)
	at org.eclipse.jface.dialogs.Dialog.create(Dialog.java:1094)
	at org.eclipse.jface.window.Window.open(Window.java:788)
	at org.eclipse.launchbar.ui.internal.LaunchBarUIManager.openConfigurationEditor(LaunchBarUIManager.java:134)
	at org.eclipse.launchbar.ui.controls.internal.ConfigSelector.handleEdit(ConfigSelector.java:156)
	at org.eclipse.launchbar.ui.controls.internal.CSelector$5.lambda$0(CSelector.java:265)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:185)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4262)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3885)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1041)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)

```

**Comment by kolipakakondal** [comment]
Maybe some issue but after delete none of the launch configurations are deleted.
Screenshot 2022-05-06 at 9 20 06 AM

After delete
Screenshot 2022-05-06 at 9 21 17 AM

**Comment by kolipakakondal** [comment]
Hi @sigmaaa Just thinking why someone would like to keep launch configurations for the deleted project, do you see any value in it? Why can't we delete all launch configurations along with the project without even asking the user? In that case, probably we don't need to do all of these, we could add a delete resource listener and if that's a project will identify all the related launch configurations and delete it. Think about it.

**Comment by kolipakakondal** [comment]
I wonder why eclipse CDT is not handling this - I was expecting this should be the default behavior of eclipse resource delete action. If the project is deleted and it should delete all the configurations associated with it. We could also introspect, is this a problem with how we create a project and launch configurations, is something can be done from outside so that eclipse can delete launch configurations along with the project

**Comment by sigmaaa** [comment]
> Hi @sigmaaa Just thinking why someone would like to keep launch configurations for the deleted project, do you see any value in it? Why can't we delete all launch configurations along with the project without even asking the user? In that case, probably we don't need to do all of these, we could add a delete resource listener and if that's a project will identify all the related launch configurations and delete it. Think about it.

Hi @kolipakakondal, 
Thanks for the review, I will fix mentioned issues. We are deleting not only launch configurations but also debug configurations, that's why I think this option should be optional.  We can use a debug configuration for another project, all we need to do is edit the debug configuration and change the project there

**Comment by kolipakakondal** [comment]
I think the root problem is with the IDFCoreLaunchConfigProvider, IDFLaunchDescriptorType and LaunchManager. It would be great if we could investigate further how we are associating launch configs with the projects, how we are persisting them and how the launch manager is handling when a project is deleted.

**Comment by sigmaaa** [comment]
Hi @kolipakakondal, @alirana01 

fixed mentioned issues with conflict handlers and creating the launch configuration. 

To create an additional launch configuration, we need to provide a project, so a new text box with the project was added to the launch configuration and now it's possible to change the project for the launch configuration. Users will see "Project not specified" or "Project does not exist" messages if a project is missing and will not be able to create a configuration without providing a project.

IEP-690

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java [hunk 1]
```diff
@@ -51,9 +67,12 @@
 import com.espressif.idf.launch.serial.SerialFlashLaunchTargetProvider;
 import com.espressif.idf.launch.serial.internal.SerialFlashLaunchConfigDelegate;
 import com.espressif.idf.launch.serial.util.ESPFlashUtil;
+import com.espressif.idf.ui.EclipseUtil;
 
+@SuppressWarnings("restriction")
 public class CMakeMainTab2 extends GenericMainTab {
 	private static final String EMPTY_CONFIG_OPTIONS = "-s ${openocd_path}/share/openocd/scripts"; //$NON-NLS-1$
+	private static final String EMPTY = ""; //$NON-NLS-1$
```
Comment (alirana01) [inline bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:75]: StringUtil.EMPTY 
can be used here from our own StringUtil class

**Review by alirana01 — COMMENTED** [review]
Code seems fine minor changes are required can you please also rebase on master I will verify the changes locally as well.

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java (hunk 1 above)
Comment (alirana01) [reply bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:75]: I updated it myself

**Review by alirana01 — APPROVED** [review]
LGTM

**Review by alirana01 — CHANGES_REQUESTED** [review]
Issue with added test case @sigmaaa please resolve it I think you faced a similar issue with tests on local machine. You may have to configure the eclipse settings in our setup class methods for initial env setup. The test gets stuck here on this screen until you press continue which means it is not expecting that in the flow. If the above settings are not an issue you can check for the continue button and if present click it or else continue with the old flow.
image

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java (hunk 1 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java]: Hi @alirana01. Thanks

File: tests/com.espressif.idf.ui.test/src/com/espressif/idf/ui/test/operations/ProjectTestOperations.java [hunk 2]
```diff
@@ -158,8 +174,18 @@ public static void deleteProject(String projectName, SWTWorkbenchBot bot, boolea
...
+			{
+				bot.checkBox("Delete all related configurations").click();
+			}
 			bot.button("OK").click();
+			if (bot.activeShell().getText().contentEquals("Delete Resources")
+					&& (bot.button(1).getText().contentEquals("Con&tinue")
+							|| bot.button(2).getText().contentEquals("Con&tinue")))
+			{
+				bot.button("Continue").click();
+			}
```
[the remark this answers is not in the document]
Comment (sigmaaa) [reply tests/com.espressif.idf.ui.test/src/com/espressif/idf/ui/test/operations/ProjectTestOperations.java]: fixed by enabling Preferences -> General -> Workspace -> Refresh using native hooks or polling in EnvSetupOperations

From eclipse documantation: 

> If this option is turned on then the workspace resources will be synchronized with their corresponding resources in the file system automatically using native refresh providers (on Windows) or a polling mechanism.

**Review by alirana01 — APPROVED** [review]
LGTM

**Pull request merged**